controllers/mirrorpeersecret_controller.go

```diff
@@ -0,0 +1,399 @@
+	client.Client
+	Scheme *runtime.Scheme
+}
+
+//+kubebuilder:rbac:groups=multicluster.odf.openshift.io,resources=mirrorpeers,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=multicluster.odf.openshift.io,resources=mirrorpeers/status,verbs=get
+//+kubebuilder:rbac:groups=core,resources=pods;secrets;configmaps;events,verbs=*
+//+kubebuilder:rbac:groups=rbac.authorization.k8s.io,resources=roles;rolebindings,verbs=*
+//+kubebuilder:rbac:groups=certificates.k8s.io,resources=certificatesigningrequests;certificatesigningrequests/approval,verbs=get;list;watch;create;update
+//+kubebuilder:rbac:groups=certificates.k8s.io,resources=signers,verbs=approve
+//+kubebuilder:rbac:groups=cluster.open-cluster-management.io,resources=managedclusters,verbs=get;list;watch
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.8.3/pkg/reconcile
+
+// Reconcile standard reconcile function for MirrorPeerSecret controller
+func (r *MirrorPeerSecretReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	// if there is no 'namespace', it must be a 'MirrorPeer' object
+	if req.Namespace == "" {
+		return r.mirrorPeerReconcile(ctx, req)
+	}
+	// all other cases, it should be 'Secret' (soucre|BLUE or destination|GREEN)
+	return r.mirrorPeerSecretReconcile(ctx, req)
+}
+
+func (r *MirrorPeerSecretReconciler) mirrorPeerReconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	logger := log.FromContext(ctx)
+
+	var mirrorPeerObj multiclusterv1alpha1.MirrorPeer
+	err := r.Get(ctx, req.NamespacedName, &mirrorPeerObj)
+	if err != nil {
+		if k8serrors.IsNotFound(err) {
+			/*
+				what all actions to be taken when a MirrorPeer CR deleted?
+			*/
+			err = nil // neglect any not found error
+		}
+		if err != nil {
+			logger.Error(err, "Error while retrieving 'MirrorPeer' object", "controller", "MirrorPeerSecret")
+		}
+		return ctrl.Result{}, err
+	}
+	logger.Info("Found a MirrorPeer Object", "controller", "MirrorPeerSecret", "MirrorPeer", mirrorPeerObj)
+
+	return ctrl.Result{}, nil
+}
+
+func (r *MirrorPeerSecretReconciler) mirrorPeerSecretReconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	logger := log.FromContext(ctx)
+	var mirrorPeerSecretObj corev1.Secret
+	err := r.Get(ctx, req.NamespacedName, &mirrorPeerSecretObj)
+	if err != nil {
+		if k8serrors.IsNotFound(err) {
+			logger.Info("Secret not found", "controller", "MirrorPeerSecret", "req", req)
+			err = r.processDeletedSecrets(ctx, req)
+		}
+		if err != nil {
+			logger.Error(err, "Error in processing deleted secrets", "controller", "MirrorPeerSecret", "request", req)
+		}
+		return ctrl.Result{}, err
+	}
+	if secretutils.IsSecretSource(&mirrorPeerSecretObj) {
+		logger.Info("It is a BLUE secret")
```

Comment:
Blue and Green are the names we are following to represent local and remote cluster secrets?